processing/src/main/java/io/druid/query/groupby/RowBasedColumnSelectorFactory.java

```diff
@@ -80,103 +85,120 @@ public Row get()
+        rowSignature
     );
   }
 
-  // This dimension selector does not have an associated lookup dictionary, which means lookup can only be done
-  // on the same row. This dimension selector is used for applying the extraction function on dimension, which
-  // requires a DimensionSelector implementation
   @Override
   public DimensionSelector makeDimensionSelector(DimensionSpec dimensionSpec)
   {
+    // This dimension selector does not have an associated lookup dictionary, which means lookup can only be done
+    // on the same row. Hence it returns CARDINALITY_UNKNOWN from getValueCardinality.
     return dimensionSpec.decorate(makeDimensionSelectorUndecorated(dimensionSpec));
   }
 
   private DimensionSelector makeDimensionSelectorUndecorated(DimensionSpec dimensionSpec)
   {
     final String dimension = dimensionSpec.getDimension();
     final ExtractionFn extractionFn = dimensionSpec.getExtractionFn();
-    final boolean isTimeColumn = dimensionSpec.getDimension().equals(Column.TIME_COLUMN_NAME);
 
-    if (isTimeColumn && extractionFn == null) {
-      throw new UnsupportedOperationException("time dimension must provide an extraction function");
-    }
+    if (Column.TIME_COLUMN_NAME.equals(dimensionSpec.getDimension())) {
+      if (extractionFn == null) {
+        throw new UnsupportedOperationException("time dimension must provide an extraction function");
+      }
 
-    return new DimensionSelector()
-    {
-      @Override
-      public IndexedInts getRow()
+      return new DimensionSelector()
       {
-        final int dimensionValuesSize;
+        @Override
+        public IndexedInts getRow()
+        {
+          return ZeroIndexedInts.instance();
+        }
 
-        if (isTimeColumn) {
-          dimensionValuesSize = 1;
-        } else {
-          final List<String> dimensionValues = row.get().getDimension(dimension);
-          dimensionValuesSize = dimensionValues != null ? dimensionValues.size() : 0;
+        @Override
+        public int getValueCardinality()
+        {
+          return DimensionSelector.CARDINALITY_UNKNOWN;
         }
 
-        return new IndexedInts()
+        @Override
+        public String lookupName(int id)
         {
-          @Override
-          public int size()
-          {
-            return dimensionValuesSize;
-          }
+          return extractionFn.apply(row.get().getTimestampFromEpoch());
+        }
 
-          @Override
-          public int get(int index)
+        @Override
+        public int lookupId(String name)
+        {
+          throw new UnsupportedOperationException("lookupId");
+        }
+      };
+    } else {
+      return new DimensionSelector()
+      {
+        @Override
+        public IndexedInts getRow()
+        {
+          final List<String> dimensionValues = row.get().getDimension(dimension);
+          final int dimensionValuesSize = dimensionValues != null ? dimensionValues.size() : 0;
+
+          return new IndexedInts()
```

Comment:
Similarly to `ZeroIndexedInts`, `RangeIndexedInts` could be extracted and certain amount of them cached globally.